File: lib/src/main/java/com/walmartlabs/components/scheduler/core/hz/ScheduleScanner.java
package com.walmartlabs.components.scheduler.core.hz;

import com.hazelcast.core.IExecutorService;
import com.hazelcast.core.IMap;
import com.hazelcast.core.Member;
import com.hazelcast.query.Predicate;
import com.walmart.gmp.ingestion.platform.framework.core.Hz;
import com.walmart.gmp.ingestion.platform.framework.core.ListenableFutureAdapter;
import com.walmart.gmp.ingestion.platform.framework.data.core.DataManager;
import com.walmart.gmp.ingestion.platform.framework.data.core.TaskExecutor;
import com.walmartlabs.components.scheduler.model.Bucket;
import com.walmartlabs.components.scheduler.model.Bucket.BucketStatus;
import com.walmartlabs.components.scheduler.services.Service;
import com.walmartlabs.components.scheduler.utils.TimeUtils;
import org.apache.log4j.Logger;
import org.springframework.beans.factory.annotation.Autowired;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.Map.Entry;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledThreadPoolExecutor;

import static com.google.common.collect.Iterators.cycle;
import static com.google.common.collect.Lists.newArrayList;
import static com.google.common.collect.Lists.partition;
import static com.google.common.collect.Maps.transformValues;
import static com.google.common.collect.Sets.newHashSet;
import static com.google.common.util.concurrent.Futures.successfulAsList;
import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
import static com.walmart.gmp.ingestion.platform.framework.core.Props.PROPS;
import static com.walmart.platform.soa.common.exception.util.ExceptionUtil.getRootCause;
import static com.walmartlabs.components.scheduler.model.Bucket.BucketStatus.PROCESSED;
import static com.walmartlabs.components.scheduler.utils.TimeUtils.toAbsolute;
import static java.lang.String.format;
import static java.time.LocalDateTime.now;
import static java.time.LocalDateTime.of;
import static java.util.concurrent.Executors.callable;
import static java.util.concurrent.TimeUnit.*;
import static java.util.stream.Collectors.toList;
import static java.util.stream.Collectors.toMap;

/**
 * Created by smalik3 on 3/8/16
 */
public class ScheduleScanner implements Service {

    static final String BUCKET_CACHE = "bucketCache";

    @Autowired
    private transient DataManager<Long, Bucket> dataManager;

    @Autowired
    private Hz hz;

    private static final ScheduledExecutorService EXECUTOR_SERVICE =
            new ScheduledThreadPoolExecutor(1, r -> new Thread(r, "EventSchedulerThread"));
    private static final Logger L = Logger.getLogger(ScheduleScanner.class);

    @Override
    public String name() {
        return "EventScanner";
    }

    @Override
    public void init() {
        L.info("initing the event scheduler");
    }

    @Override
    public void destroy() {
        L.info("destroying the event scheduler");
    }

    @Override
    public void execute() {
        if (true)
            return;
        L.info("executing the EventScheduleScanner");
        final Integer scanInterval = PROPS.getInteger("event.schedule.scan.interval.minutes", 60);
        L.info("calculating the next scan bucketId");
        final LocalDateTime now = now();
        final LocalDateTime nextScan = TimeUtils.nextScan(now, scanInterval);
        final long delay = ChronoUnit.MILLIS.between(now, nextScan);
        L.info(format("first-scan at: %s, next-scan at: %s, " +
                "initial-delay: %d ms, subsequent-scans: after every %d minutes", now(), nextScan, delay, scanInterval));
        EXECUTOR_SERVICE.scheduleAtFixedRate(this::scan, delay, MILLISECONDS.convert(scanInterval, MINUTES), MILLISECONDS);
        L.info("executing first time scan");
        scan();
    }

    private void scan() {
        L.debug("scanning the schedule(s)");
        try {
            final ChronoUnit chronoUnit = ChronoUnit.valueOf(PROPS.getProperty("event.schedule.scan.unit", ChronoUnit.HOURS.name()));
            final long currentBucketId = chronoUnit.between(of(now().getYear(), 1, 1, 0, 0), now());
            final int lookbackRange = PROPS.getInteger("events.backlog.check.limit", 2);
            final Set<Long> bucketIds = new HashSet<>();
            for (long i = currentBucketId - lookbackRange; i <= currentBucketId; i++) {
                bucketIds.add(i);
            }
            final Map<Long, Set<Integer>> shards = calculateScheduleDistribution(currentBucketId, bucketIds);

            if (shards.isEmpty()) {
                L.debug(format("%d, nothing to schedule for bucketIds: %s", currentBucketId, bucketIds));
                return;
            }
            if (!bucketIds.isEmpty())
                L.debug(format("%d, following bucketIds will not be scheduled, (either no events or all processed): %s", currentBucketId, bucketIds));

            final int submitRetries = PROPS.getInteger("event.submit.max.retries", 10);
            final int submitInitialDelay = PROPS.getInteger("event.submit.initial.delay", 1);
            final int submitBackoff = PROPS.getInteger("event.submit.backoff.multiplier", 1);

            L.debug("submitting the schedules for execution" + shards);
            final IExecutorService executorService = hz.hz().getExecutorService("default");
            final Set<Member> members = hz.hz().getCluster().getMembers();
            final Iterator<Member> iterator = cycle(members);

            successfulAsList(partition(newArrayList(shards.entrySet()), members.size() != 1 ? shards.size() / members.size() : 1).stream().
                    map(l -> l.stream().collect(toMap(Entry::getKey, Entry::getValue))).
                    collect(toList()).stream().map(BulkEventTask::new).collect(toList()).stream().map(e ->
                    taskExecutor.async(() -> () -> ListenableFutureAdapter.<Map<Long, BucketStatus>>adapt(
                            executorService.submitToMember(callable(e), iterator.next())), "event-submit",
                            submitRetries, submitInitialDelay, submitBackoff, SECONDS)).collect(toList())).addListener(() ->
                    L.info("shards submitted successfully for processing: " + shards), directExecutor());
        } catch (Exception e) {
            L.error("schedule scan failed", getRootCause(e));
        }
    }

    private final TaskExecutor taskExecutor = new TaskExecutor(newHashSet(Exception.class));

    private Map<Long, Set<Integer>> calculateScheduleDistribution(long calcId, final Set<Long> bucketIds) {
        final IMap<Long, Bucket> cache = hz.hz().getMap(BUCKET_CACHE);
        L.debug(format("%d, checking if following bucketIds have schedules: %s", calcId, bucketIds));
        cache.getAll(bucketIds);
        return transformValues(cache.executeOnEntries(null,
                new Predicate<Long, Bucket>() {
                    @Override
                    public boolean apply(Entry<Long, Bucket> e) {
                        return bucketIds.contains(e.getKey()) && e.getValue() != null && e.getValue().getCount() > 0
                                && !PROCESSED.name().equals(e.getValue().getStatus());
                    }
                }), $ -> (Bucket) $).entrySet().stream().collect(toMap(Entry::getKey, e -> {
            final Long bucketId = e.getKey();
            final Bucket bucket = e.getValue();
            L.debug(format("%d, scheduling bucket: %d", calcId, bucketId));
            bucketIds.remove(bucketId);
            L.debug(format("%d, calculating schedule distribution for bucketId: %d", calcId, bucketId));
            int shardSize = PROPS.getInteger("event.shard.size", 1000);
            final int numShards = (int) bucket.getCount() / shardSize + 1;
            L.debug(format("%d, bucketId: %d, numShards: %d, count: %d, shardSize: %d", calcId, bucketId, numShards, bucket.getCount(), shardSize));
            L.debug(format("%d, submitting the schedules over the grid", calcId));
            final List<Integer> shardIndexes = new ArrayList<>();
            for (int i = 0; i < numShards; i++) {
                shardIndexes.add(i);
            }
            final int gridSize = hz.hz().getCluster().getMembers().size();
            final List<List<Integer>> partitions = partition(shardIndexes, numShards < gridSize ? 1 : numShards / gridSize);
            final long bucketKey = toAbsolute(bucketId);
            L.debug(format("%d, distribution profile: bucketKey: %d, nodes: %d, partitions: %s", calcId, bucketKey, gridSize, partitions));
            return newHashSet(shardIndexes);
        }));
    }
}

File: lib/src/main/kotlin/com/walmartlabs/bigben/core/ScheduleScanner.kt
/*-
 * #%L
 * BigBen:lib
 * =======================================
 * Copyright (C) 2016 - 2018 Walmart Inc.
 * =======================================
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package com.walmartlabs.bigben.core

import com.google.common.base.Throwables.getRootCause
import com.google.common.collect.Iterators
import com.google.common.collect.LinkedHashMultimap
import com.google.common.collect.Multimap
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.MoreExecutors.listeningDecorator
import com.hazelcast.core.IExecutorService
import com.hazelcast.core.Member
import com.walmartlabs.bigben.entities.EventStatus.ERROR
import com.walmartlabs.bigben.entities.EventStatus.PROCESSED
import com.walmartlabs.bigben.entities.ShardStatus
import com.walmartlabs.bigben.entities.ShardStatusList
import com.walmartlabs.bigben.extns.bucketize
import com.walmartlabs.bigben.extns.nextScan
import com.walmartlabs.bigben.extns.nowUTC
import com.walmartlabs.bigben.tasks.BulkShardTask
import com.walmartlabs.bigben.utils.*
import com.walmartlabs.bigben.utils.hz.Hz
import com.walmartlabs.bigben.utils.hz.Service
import com.walmartlabs.bigben.utils.utils.Props
import com.walmartlabs.bigben.utils.utils.Props.int
import com.walmartlabs.bigben.utils.utils.Props.string
import java.lang.Runtime.getRuntime
import java.time.Instant
import java.time.ZoneOffset.UTC
import java.time.ZonedDateTime
import java.time.temporal.ChronoUnit
import java.util.concurrent.Executors.newFixedThreadPool
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit.*
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicReference

/**
 * Created by smalik3 on 2/23/18
 */
class ScheduleScanner(private val hz: Hz) : Service {

    companion object {
        internal const val BUCKET_CACHE = "bucketCache"
        const val EVENT_SCHEDULER = "event_scheduler"

        private val l = logger<ScheduleScanner>()

        private val index = AtomicInteger()
        private val shardIndexer = AtomicInteger()
        private val scheduler = listeningDecorator(ScheduledThreadPoolExecutor(getRuntime().availableProcessors()) { r -> Thread(r, "InternalScheduler#" + index.getAndIncrement()) })
        private val shardSubmitters = listeningDecorator(newFixedThreadPool(int("events.num.shard.submitters")) { r -> Thread(r, "ShardSubmitter#" + shardIndexer.getAndIncrement()) })
    }

    private val isShutdown = AtomicReference(false)

    private lateinit var bucketManager: BucketManager
    private val bucketWidth = int("events.schedule.scan.interval.minutes")
    @Volatile
    private lateinit var lastScan: ZonedDateTime

    override val name: String = "ScheduleScanner"

    override fun init() {
        if (l.isInfoEnabled) l.info("initing the event scheduler")
        val lookbackRange = int("buckets.backlog.check.limit")
        val checkpointInterval = Props.long("buckets.checkpoint.interval")
        val checkpointIntervalUnits = valueOf(string("buckets.checkpoint.interval.units"))
        bucketManager = BucketManager(lookbackRange + 1, 2 * bucketWidth * 60, bucketWidth * 60, checkpointInterval, checkpointIntervalUnits, hz)
    }

    override fun execute() {
        if (l.isInfoEnabled) l.info("executing the EventScheduleScanner")
        val scanInterval = int("events.schedule.scan.interval.minutes")
        if (l.isInfoEnabled) l.info("calculating the next scan bucketId")
        val now = nowUTC()
        val nextScan = nextScan(now, scanInterval)
        val delay = ChronoUnit.MILLIS.between(now, nextScan)
        val bucket = ZonedDateTime.ofInstant(Instant.ofEpochMilli(bucketize(now.toInstant().toEpochMilli(), scanInterval)), UTC)
        lastScan = bucket.minusMinutes(bucketWidth.toLong())
        if (l.isInfoEnabled) l.info("first-scan at: {}, for bucket: {}, next-scan at: {}, " + "initial-delay: {} ms, subsequent-scans: after every {} minutes", now, bucket, nextScan, delay, scanInterval)
        scheduler.scheduleAtFixedRate({ this.scan() }, delay, MILLISECONDS.convert(scanInterval.toLong(), MINUTES), MILLISECONDS)
        if (l.isInfoEnabled) l.info("executing first time scan")
        scan()
    }

    private fun scan() {
        if (isShutdown.get()) {
            if (l.isInfoEnabled) l.info("system is shutdown, no more schedules will be processed")
            return
        }
        val currentBucketId = lastScan.plusMinutes(bucketWidth.toLong())
        lastScan = currentBucketId
        if (l.isDebugEnabled) l.debug("scanning the schedule(s) for bucket: {}", currentBucketId)
        try {
            bucketManager.getProcessableShardsForOrBefore(currentBucketId).done({ l.error("error in processing bucket: {}", currentBucketId, it!!.rootCause()) }) {
                try {
                    if (it!!.isEmpty) {
                        if (l.isDebugEnabled) l.debug("nothing to schedule for bucket: {}", currentBucketId)
                        return@done
                    }
                    shardSubmitters.submit {
                        if (l.isDebugEnabled) l.debug("{}, shards to be processed: => {}", currentBucketId, it)
                        calculateDistro(it).asMap().run {
                            if (l.isInfoEnabled) l.info("{}, schedule distribution: => {}", currentBucketId,
                                    mapKeys { it.key.address.toString() }.mapValues { it.value.joinToString(",") { "${it.first}[${it.second}]" } }.toSortedMap())

                            val iterator = Iterators.cycle<Member>(keys)
                            val executorService = hz.hz.getExecutorService(EVENT_SCHEDULER)
                            entries.map {
                                { submitShards(executorService, iterator.next(), it.value, currentBucketId) }.retriable("shards-submit", int("event.submit.max.retries"),
                                        int("events.submit.initial.delay"),
                                        int("events.submit.backoff.multiplier")).transform { it!!.list }
                            }.reduce().done({ l.error("schedule for bucket {} finished abnormally", currentBucketId, it.rootCause()) }) {
                                if (l.isDebugEnabled) l.debug("schedule for bucket {} finished normally => {}", currentBucketId, it)
                                val buckets = it!!.map { it!! }.flatten().filterNotNull().groupBy { it.bucketId!! }.mapValues { it.value.fold(false) { hasError, ss -> hasError || (ss.status == ERROR) } }
                                if (l.isDebugEnabled) l.debug("bucket-scan: {}, final buckets with statuses to be persisted: {}", currentBucketId, buckets)
                                buckets.map { bucketManager.bucketProcessed(it.key, if (it.value) ERROR else PROCESSED) }.done({ l.error("bucket-scan: {}, failed to update the scan-status: {}", currentBucketId, buckets.keys, it.rootCause()) }) {
                                    if (l.isDebugEnabled) l.debug("bucket-scan: {}, successfully updated the scan-status: {}", currentBucketId, buckets.keys)
                                }
                            }
                        }
                    }
                } catch (e: Exception) {
                    l.error("error in processing bucket: {}", currentBucketId, getRootCause(e))
                }
            }
        } catch (e: Exception) {
            l.error("error in processing bucket: {}", currentBucketId, getRootCause(e))
        }
    }

    private fun calculateDistro(shards: Multimap<ZonedDateTime, Int>): LinkedHashMultimap<Member, Pair<ZonedDateTime, Int>> {
        val members = hz.hz.cluster.members.toMutableSet().apply { remove(hz.hz.cluster.localMember) }.toList().shuffled().toMutableList().apply { add(hz.hz.cluster.localMember) }
        val entries = shards.entries().toList()
        return LinkedHashMultimap.create<Member, Pair<ZonedDateTime, Int>>().apply {
            val size = members.size
            for (i in entries.indices) {
                val e = entries[i]
                put(members[i % size], e.key to e.value)
            }
        }
    }

    private fun submitShards(executorService: IExecutorService, member: Member, shardsData: Collection<Pair<ZonedDateTime, Int>>, bucket: ZonedDateTime): ListenableFuture<ShardStatusList> {
        if (l.isDebugEnabled) l.debug("{}, submitting  for execution to member {}, shards: {}", bucket, member.socketAddress, shardsData)
        bucketManager.registerForProcessing(shardsData)
        return executorService.submitToMember(BulkShardTask(shardsData), member).listenable().catching { ShardStatusList(shardsData.map { ShardStatus(it.first, it.second, ERROR) }) }.done({
            l.error("{}, member {} finished abnormally for shards: {}", bucket, member.socketAddress, shardsData, it)
            shardsData.forEach { bucketManager.shardDone(it.first, it.second, ERROR) }
        }) {
            if (l.isDebugEnabled) l.debug("{}, member {} finished normally for shards: {}", bucket, member.socketAddress, it)
            it?.list?.forEach { bucketManager.shardDone(it!!.bucketId!!, it.shard!!, it.status!!) }
        }
    }

    override fun destroy() {
        if (l.isInfoEnabled) l.info("destroying the event scheduler")
    }
}
